### PR TITLE
Add discard reason in eth_getTransactionReceipt

### PR DESCRIPTION
**Pull request by Kourin1996**
# Description

This PR lets `eth_getTransactionReceipt` return message indicating the reason of tx discard.

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "eth_getTransactionReceipt",
  "params": [
    "0x...." 
  ]
}
```

If the reason is recorded in journal, JSON-RPC returns string reason as an error instead of nil.

```json
{
  "message": "Tx discarded by reaching recover limit"
}
```

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

1. Setup & Start Edge clusters in the usual manner
2. Send a transaction
3. Call `eth_txStatus` giving the tx hash and make sure the status is right

# Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: txpool/journal.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+// memoryJournal is an implementation of Journal and it saves data in-memory
+type memoryJournal struct {
+	txHashMap sync.Map
+}
+
+// newMemoryJournal initializes memoryJournal
+func newMemoryJournal() *memoryJournal {
+	return &memoryJournal{
+		txHashMap: sync.Map{},
```
Comment (ferranbt) [inline txpool/journal.go]: Does the map implement a cache? Otherwise, a malicious actor could attack the pool sending thousands of dropped transactions.

File: e2e/framework/config.go [hunk 2]
```diff
@@ -46,6 +46,7 @@ type TestServerConfig struct {
 	ValidatorType           validators.ValidatorType // Validator Type
 	Bootnodes               []string                 // Bootnode Addresses
 	PriceLimit              *uint64                  // Minimum gas price limit to enforce for acceptance into the pool
+	MaxSlots                *uint64                  // Maximum available slots in TxPool
```
Comment (ferranbt) [inline e2e/framework/config.go:49]: Could we move this to another PR or is required for this?

File: jsonrpc/eth_endpoint.go [hunk 3]
```diff
@@ -769,6 +775,11 @@ func (e *Eth) Unsubscribe(id string) (bool, error) {
 	return e.filterManager.Uninstall(id), nil
 }
 
+// TxStatus returns a status of the Tx that has been added in TxPool
+func (e *Eth) TxStatus(txHash types.Hash) (*string, error) {
```
Comment (ferranbt) [inline jsonrpc/eth_endpoint.go]: Could this be merged with `eth_getTransactionReceipt`?

File: txpool/journal.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+const (
+	TxSuccessful TxStatus = iota // Tx has been mined in a block successfully
+	TxDropped                    // Tx was dropped for some reason
+)
+
+var (
+	// txStatusToString is a mapping from TxStatus to text status
+	txStatusToString = map[TxStatus]string{
+		TxSuccessful: "success",
```
Comment (ferranbt) [inline txpool/journal.go]: I think the journal should be used only to signal when a transaction was dropped from the pool and why.

File: txpool/journal.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+		TxDropped:    "dropped",
```
Comment (ferranbt) [inline txpool/journal.go]: Dropped should have a bit more context on why was it dropped: gas limit, not enough space in pool, nonce gap too big...

**Review by ferranbt — CHANGES_REQUESTED** [review]

File: txpool/journal.go (hunk 1 above)
Comment (Kourin1996) [reply txpool/journal.go]: it holds tx statuse like mined （processed in a block already）or dropped.
Yes, you're right. Someone can send many txs that can't be processed （unconsective nonces） and occupy node's memory.
I think tx has 4 statuses
unconfirmed: tx doesn't reach to node
pending: tx is in TxPool
dropped: tx has been removed from TxPool some reason (unrecoverable error, recoverable error but error happened many times, and cleaning up nonce holes)
successful: tx has been processed and packed in a block

You can get tx with pending or successful status by existing API. But you can't get dropped or unconfirmed tx. That's why I added the dropped tx's hash in journal now

Do you think we don't need record dropped tx? or need cleanup function to keep memory usage stable?

File: e2e/framework/config.go (hunk 2 above)
Comment (Kourin1996) [reply e2e/framework/config.go:49]: It's used in new e2e test to drop tx from tx pool due to nonce hole cleanups. It's easy and simple way. but let me try to think about another approach

File: jsonrpc/eth_endpoint.go (hunk 3 above)
Comment (Kourin1996) [reply jsonrpc/eth_endpoint.go]: So eth_getTransactionReceipt will return the following types of value?
・receipt body
・"success"
・"dropped"
・null

File: txpool/journal.go (hunk 4 above)
Comment (Kourin1996) [reply txpool/journal.go]: it can be after we merge with 'eth_getTransactionReceipt'

File: txpool/journal.go (hunk 5 above)
Comment (Kourin1996) [reply txpool/journal.go]: Do we need to return this reason as a part of JSONRPC response? 😄

File: e2e/framework/config.go (hunk 2 above)
Comment (lazartravica) [reply e2e/framework/config.go:49]: This is only for controlling the txpool max memory footprint in the e2e tests, it's not a MaxSlot implementation as that is already present in the txpool just hasn't been used in tests before @ferranbt @Kourin1996

File: txpool/journal.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+// newMemoryJournal initializes memoryJournal
+func newMemoryJournal() *memoryJournal {
+	return &memoryJournal{
+		txHashMap: sync.Map{},
+	}
+}
+
+// logSuccessfulTx marked the Tx as mined
+func (j *memoryJournal) logSuccessfulTx(txHash types.Hash) {
+	j.txHashMap.Store(txHash, true)
```
Comment (lazartravica) [inline txpool/journal.go]: We don't need to log successful txs.

**Review by lazartravica — CHANGES_REQUESTED** [review]
I don't think that we need journaling of successful transactions.

The only purpose of this new journal is to keep track of the transactions that are acceptable to the TxPool, but were dropped from it (for a multitude of reasons).

To summarize, a couple of big changes that need to be done on the PR:
- Log only dropped transactions from TxPool (ones that are rejected from JSON-RPC we don't care about, and ones that are included in a block there is the getTransactionReceipt call)
- The journal should be bounded in memory, something like a circular list where the eleements would overwrite in chronological order would be perfect (ideally without allocating memory all at once, imo).
- Add descriptive reasons (there's a finite number of them to why it got dropped).

As for reusing the `eth_getTransactionReceipt` or the `eth_getTransactionByHash` endpoint with these informations, I suggest to think about that later, not to start breaking web3 clients on accident and just provide this endpoint.

Also, @Kourin1996 we need a linked Docs PR for this endpoint. There's no docs on this other than the ones we write :) 

@ferranbt @Kourin1996 please let's reach consensus first here in the discussion , before going into implementation 🙏

**Comment by Kourin1996** [comment]
@lazartravica 
> The journal should be bounded in memory, something like a circular list where the eleements would overwrite in chronological order would be perfect (ideally without allocating memory all at once, imo)

I have no idea of the best data structure but journal can have ring buffer (circular list) to hold keys and lookup map from key to data (flag and reason,etc)

**Comment by ferranbt** [comment]
Agree with everything that @lazartravica mentions, some of my reviews go in that direction too.

> As for reusing the eth_getTransactionReceipt or the eth_getTransactionByHash endpoint with these informations, I suggest to think about that later, not to start breaking web3 clients on accident and just provide this endpoint.

I think it is better to update `eth_getTransactionReceipt` since that one represents the end status of a transaction.

I am not worried about breaking the endpoint, this is just an error message (if as discussed we only journal the drop reasons). I think it is way more cumbersome to add support for new endpoints in web3 clients (I am not sure I would add it on `ethgo` only to give support for `edge` for example).

> I have no idea of the best data structure but journal can have ring buffer (circular list) to hold keys and lookup map from key to data (flag and reason,etc)

Use a simple cache like https://github.com/hashicorp/golang-lru. Lets not overcomplicate the design.

**Comment by Kourin1996** [comment]
@ferranbt Thank you for reviewing the discussion!
cc: @lazartravica 

> Use a simple cache like https://github.com/hashicorp/golang-lru. Lets not overcomplicate the design

In such a case, I thought the data in journal wouldn't be discarded in FIFO. Is it ok for you?

File: txpool/journal.go (hunk 1 above)
Comment (JekaMas) [reply txpool/journal.go]: > Does the map implement a cache? Otherwise, a malicious actor could attack the pool sending thousands of dropped transactions.

Had exactly the same question in the RFC :)

It might make sense to have cache based also on accounts or have the bigger TTL the greater nonce is.

File: jsonrpc/eth_endpoint.go (hunk 3 above)
Comment (JekaMas) [reply jsonrpc/eth_endpoint.go]: I don't think that it is related to receipts. A transaction might not have a receipt at all.

File: txpool/journal.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	"sync"
+
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+// TxStatus indicates the status of tx that is added to TxPool
+type TxStatus int8
+
+const (
+	TxSuccessful TxStatus = iota // Tx has been mined in a block successfully
```
Comment (JekaMas) [inline txpool/journal.go]: Why we need this statuses as ints, not strings?

Comment (JekaMas) [reply txpool/journal.go]: Is it for performance?

Comment (JekaMas) [reply txpool/journal.go]: if so then it's better to start with initial value 1, not 0 and treat 0 as incorrect.

File: txpool/journal.go (hunk 5 above)
Comment (JekaMas) [reply txpool/journal.go]: "gas limit, not enough space in pool, nonce gap too big" we might need types errors for users with error codes as well. what do you think?

File: txpool/journal.go (hunk 6 above)
Comment (JekaMas) [reply txpool/journal.go]: Do we define 'slot' the same way as in Geth?

Comment (lazartravica) [reply txpool/journal.go]: @JekaMas yep, https://github.com/0xPolygon/polygon-edge/blob/b97e753e8300e488e88a9046bf2aa2628bd269cc/txpool/txpool.go#L24-L25

**Comment by lazartravica** [comment]
> @ferranbt Thank you for reviewing the discussion! cc: @lazartravica
> 
> > Use a simple cache like https://github.com/hashicorp/golang-lru. Lets not overcomplicate the design
> 
> In such a case, I thought the data in journal wouldn't be discarded in FIFO. Is it ok for you?

I agree completely with @ferranbt suggestion let's not introduce complexity and reuse the package we already have, the LRU cache itself will prove to be good.

I am also in-line with `eth_getTransactionReceipt`, though not aware of the complexity of making that change.

cc @Kourin1996

File: jsonrpc/eth_endpoint.go [hunk 8]
```diff
@@ -295,7 +298,20 @@ func (e *Eth) GetTransactionByHash(hash types.Hash) (interface{}, error) {
...
+		if err != nil {
+			e.logger.Warn(
+				"failed to get the reason of tx discard: hash=%s, err=%s",
+				hash,
+				err.Error(),
+			)
+		}
+
+		if reason != nil {
+			return *reason, nil
```
Comment (ferranbt) [inline jsonrpc/eth_endpoint.go]: This is not API-compatible with the JsonRPC spec, it might cause problems with the client implementation. I was under the impression that we would return it as an error of the JsonRPC call, which would be more compatible, and it kind of makes sense since this is an error on the transaction.

**Review by ferranbt — CHANGES_REQUESTED** [review]

File: jsonrpc/eth_endpoint.go (hunk 8 above)
Comment (Kourin1996) [reply jsonrpc/eth_endpoint.go]: Fixed 05377c9d21c5dc0241a8c9637eafdfc17b0b0bc9
